### ISSUE TITLE
Routing Spec performance issues in 7.0?

### ISSUE THREAD

**Issue by agilmore2** [opendcs/opendcs#255]
I'm porting a use case from 6.8 to 7.0.8, and finding routing specs are taking a very long time to run.

In 6.8-RC15, a test loads 428 messages in 3 minutes 41 seconds. After 3 hours 36 minutes, 7.0.8 only loaded 277 messages.

What's going on?

**Desktop/ServerSide (please complete the following information):**
 - OS: RHEL7.9 
 - Java: OpenJDK 8_372

**Comment by agilmore2** [comment]
[7.0.8 log](https://github.com/opendcs/opendcs/files/11994348/d0test.log)
[6.8-RC15 log](https://github.com/opendcs/opendcs/files/11994349/test.log)

**Comment by agilmore2** [comment]
[7.0.8 -d 3 log](https://github.com/opendcs/opendcs/files/11994422/d3test.log)
File collected earlier than the other two, still attempting a 3 hour DDS load.

**Comment by MikeNeilson** [comment]
If the messages are all GOES based can you create a DB export of the routingspec and network list? 
We've been having issues with the scheduler just not working but so far it's been fast when it does.

But some changes were made to handle the connection pool and a few of the DAOs will do a database query each creation. Can you run the routing spec with Java Flight Recorder, or since it's clearly taking so long, with VisualVM running and sampling the CPU time? I can't think of anything in particular off the top of my head but that instrumentation should make it really obvious.

**Comment by MikeNeilson** [comment]
Looking through the log though I'm seeing several HDB interactions that are taking longer than I would expect, and would certainly add up.

The first was
```
INFO    07/08/23 21:25:23 HdbConsumer 0090432C Looking up TSID for '2616.65.instant.R_'
DBG3    07/08/23 21:25:26 HdbTimeSeriesDAO getTimeSeriesIdentifier for '2616.65.instant.R_'
```

And that's pretty consistent throughout. Depending on how many stations are in here that could really add up.

**Comment by MikeNeilson** [comment]
Hmm, it's taking 3 seconds to go from The first log message to the next log message. There's only a function call in between.

I can see that happening with CWMS because part of the pool was messed up, but the above instrumentation would be required to know why it's going on with HDB.

**Comment by agilmore2** [comment]
My first guess is that the number of sites is an issue, there's only 133ish
platforms in the netlists, but there's over 5000 sites in the database.
I'll collect what you suggested to see.

Thanks for the weekend responses!

**Comment by MikeNeilson** [comment]
I'd also try to look at filesystem performance. 3 seconds between
consectutive log message is beyond excessive.

**Comment by agilmore2** [comment]
The 6.8 tests were run on exactly the same system after the 7.0.8 test and
before the 7.0.4 and 7.0.1.

I didn't get performance numbers off of the earlier versions, but they
weren't much different.

I see site cache refreshes as occurring much more often.

**Comment by MikeNeilson** [comment]
Fair point. Could be the dacq event logger. that writes to the database but due to the way the log works it can't actually write it's own log messages. Definitely one of those areas where micrometer would come in handy. Well, and just visualvm.

**Comment by agilmore2** [comment]
server is headless with only ssh to it, going to need to track down how to
instrument this

**Comment by MikeNeilson** [comment]
VisualVM can do remote connections (I'd tunnel through SSH though) but you'll have to add some extra parameters to the JVM for it to work. 

I've found adding things to DECJ_MAXHEAP generally just works. It just gets expanded into the java command as is.

**Comment by MikeNeilson** [comment]
https://docs.oracle.com/javase/8/docs/technotes/guides/visualvm/applications_remote.html

or you just need "jstatd"

**Comment by agilmore2** [comment]
[dbexport.log](https://github.com/opendcs/opendcs/files/12011452/dbexport.log)
dbexport. Not entirely sure why it's showing only 371 sites, when dbedit clearly shows many more.

**Comment by MikeNeilson** [comment]
oh, can you upload that dbexport (but stripped of creds) again. I'm actually at the point of my homelab were I need data and that'll be easier than hunt for info from CDEC... though I do need to do that also.

**Comment by MikeNeilson** [comment]
huh, the only things I'm seeing that are clickable are the logs.

**Comment by MikeNeilson** [comment]
oh never mind it was the "log".

**Comment by agilmore2** [comment]
GitHub wouldn't let me attach a .xml file iirc

**Comment by MikeNeilson** [comment]
Weird, but what isn’t these days.

From: agilmore2 ***@***.***>
Sent: Thursday, August 17, 2023 9:04 AM
To: opendcs/opendcs ***@***.***>
Cc: Mike Neilson ***@***.***>; Comment ***@***.***>
Subject: [URL Verdict: Neutral][Non-DoD Source] Re: [opendcs/opendcs] Routing Spec performance issues in 7.0? (Issue #255)

GitHub wouldn't let me attach a .xml file iirc

**Comment by MikeNeilson** [comment]
i finally got my test environment to the point were i could check on this. the provided export's ALL_UC_Sites ran in 42.3 real seconds to the reference postgres database.

**Comment by agilmore2** [comment]
Thank you! What's the count of the Sites table?

**Comment by MikeNeilson** [comment]
344, so perhaps not the fairest of tests.

**Comment by agilmore2** [comment]
The source database contains 5700ish sites iirc.

**Comment by agilmore2** [comment]
Just tested with an RDS version of the source database, both 6.8rc15 and 7.0.10 took 41 minutes to process a day (3400+ GOES messages) for a realtime routing spec with a local client and the database in AWS. The production database takes 2 hours 15 minutes for 7.0.10 and 5 minutes for 6.8RC15 when running to a database on localhost for the same amount of messages.

I'm really confused as to how 7.0.10 and 6.8rc15 can act the same when connecting from my home servers to AWS RDS, but production acts very differently connecting locally. One database is a clone of the other.

The decodes.properties/user.properties files are the same between versions.

It's not the ojdbc6 vs ojdbc8 difference either, I tested 7.0.10 with ojdbc6, and no difference in time.

Tested routing specs on the same network as the AWS RDS, and both 6.8 and 7.0.10 finished in 3 minutes.

**Comment by MikeNeilson** [comment]
One of the next questions is "what else is going on in the database."
One of the USACE staff has created some really good documentation on reviewing various oracle measuring APIs. When I'm in the office tomorrow I'll get you an account so you review them. But off the top of my head, `v$session` and `v$sqlarea?` and the ADDM report (probably the easiest to look at) could give you some insight.

41 minutes is a while, but given that difference I strongly suspect more than just OpenDCS is avoiding bind variables talking to that database; that'll cause query plans to be kick out of the `v$sqlarea` and plans need to be rebuild too often; basically a compilation so it can take a while.

**Comment by MikeNeilson** [comment]
wait. For the prod database, was the 6.8 RoutingScheduler *AND* 7.0 RoutingScheduler both running at the same time?
Because if so I could see the general lack of bind vars playing hell with the query plans. Not sure it would do 41 minutes to 2.25 hours though.

**Comment by agilmore2** [comment]
None of these were concurrent.

The 41 minute runs are remote network to an AWS clone of the database. Both
68 and 70 showed same timings.

The change is 68 in 5 minutes locally to the production system, and 2:15
with 70 in the same circumstances.

Local to the AWS Database was 3 minutes each!

Very strange!

**Comment by MikeNeilson** [comment]
Yeah that's crazy. Sounds like Java Flight Recorder is the next thing. Should identify hot spots pretty well.
Actually even Jconsole or visualvm after 20 minutes or show should give you decent idea. But JFR is better for that fire and forget type testing.

**Comment by agilmore2** [comment]
Have JFR from 6.8 and 7.0 runs. Looks like it's all Socket IO delays from
Oracle. Why would it be only getting delays from 7 in one database, and not
from export/reimports of the same database?!?

**Comment by MikeNeilson** [comment]
@agilmore2 have you tried running these again recently?
